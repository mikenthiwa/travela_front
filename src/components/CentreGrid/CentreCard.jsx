import React, { PureComponent} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import FemaleIcon from '../../images/icons/Female-grey.svg';
import MaleIcon from '../../images/icons/Male-grey.svg';
import UnisexIcon from '../../images/icons/Unisex-grey.svg';
import Button from '../buttons/Buttons';
import bedIcon from '../../images/icons/accomodation-grey.svg';
import bathroomIcon from '../../images/icons/hot_tub_24px.svg';
import defaultImage from '../../images/default-image.jpg';
import Utils from '../../helper/Utils';
import './CentreCard.scss';

class CentreCard extends PureComponent {

  static propTypes = {
    cardImage: PropTypes.string.isRequired,
    imageAlt: PropTypes.string.isRequired,
    countryFlagImage: PropTypes.string.isRequired,
    guestHouseName: PropTypes.string.isRequired,
    guestHouseLocation: PropTypes.string.isRequired,
    beds: PropTypes.number.isRequired,
    bathrooms: PropTypes.number.isRequired,
    guesthouseId: PropTypes.string.isRequired,
    disabledGuestHouse: PropTypes.object.isRequired,
    handleOnRestore: PropTypes.func,
    guestHouse: PropTypes.object.isRequired
  };
  static defaultProps = {
    handleOnRestore: () => null
  }
  

  genderPolicy = {
    Unisex: UnisexIcon,
    Female: FemaleIcon,
    Male: MaleIcon
  }
  
  handleClick(e) {
    e.preventDefault();
  }
  
  renderCentreFacilities(icon, facility, facilityNumber) {
    return (
      <div className="centre__icon-container">
        <img src={icon} alt={`${facility} number`} className="centre__icon" />
        <span className="centre__facility-number">{facilityNumber}</span>
        {Utils.formatWord(facility, facilityNumber)}
      </div>
    );
  }

  renderCentreImage(image, imageAlt) {
    return (
      <div className="mdl-card__media centre-card__image">
        <img
          src={image}
          alt={imageAlt}
          className="centre__image"
          onError={(e) => e.target.src = defaultImage} />
      </div>
    );
  }

  renderCentreFlag(guestHouseLocation, flagImage) {
    return (
      <div
        className="centre__flag"
        style={{backgroundImage: `url(${flagImage})`}}
        alt={`${guestHouseLocation} flag`}
      />
    );
  }

  renderCentreInfo(houseName, location) {
    return (
      <div>
        <p className="centre__name">
          {houseName}
        </p>
        <span className="centre__location">{location}</span>
      </div>
    );
  }

  renderRestoreButton() {
    const { handleOnRestore, guesthouseId } = this.props;
    return (
      <Button
        buttonClass="restore-acc-btn"
        onClick={() => handleOnRestore(guesthouseId)}
        text="Restore"
      />
    );
  }

  render() {
    const { cardImage, imageAlt, countryFlagImage, guestHouseName, guestHouseLocation, beds,
      bathrooms, guesthouseId, disabledGuestHouse, guestHouse } = this.props;
    return (
      <div className={disabledGuestHouse.id ? 
        'mdl-cell mdl-cell--4 mdl-card centre-card-disabled' 
        : 'mdl-cell mdl-cell--4 mdl-card centre-card'}>
        <Link
          to={`/residence/manage/guest-houses/${guesthouseId}`} 
          onClick={disabledGuestHouse.id ? this.handleClick : ()=>null}
          id="thisGuesthouseLink"
        >
          { this.renderCentreImage(cardImage, imageAlt)}
          { disabledGuestHouse.id ? this.renderRestoreButton() : '' }
          <div className="mdl-card__supporting-text centre-info">
            <div className="centre-info__container">
              {this.renderCentreFlag(guestHouseLocation, countryFlagImage)}
              {this.renderCentreInfo(guestHouseName, guestHouseLocation)}
            </div>
          </div>
          <div className="mdl-card__actions centre__description">
            <div className="centre__horizontal-line" />
            <div className="centre__facility">
              {this.renderCentreFacilities(bedIcon, 'Bed', beds)}
              <img src={`${this.genderPolicy[guestHouse.genderPolicy]}`} alt={guestHouse.genderPolicy} title={guestHouse.genderPolicy} />
              {this.renderCentreFacilities(bathroomIcon, 'Bathroom', bathrooms)}
            </div>
          </div>
        </Link>
      </div>
    );
  }
}

export default CentreCard;
