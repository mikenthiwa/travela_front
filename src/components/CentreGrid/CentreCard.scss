$white-color: #FFF;
$text-highlight-color: #4F4F4F;
$outline-color: #E4E4E4;
$text-color: #999;
$normal-font: "DIN Pro";
$bold-font: "DIN Pro Medium";

.centre {
  &-card {
    margin: 8px 15px 8px 0;
    &__image {
      height: 200px;
      background-color: #DCDCDC;
    }
  }
  &-card-disabled {
    margin: 8px 15px 8px 0;
    &__image {
      height: 200px;
    }
    &:before {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      background-image: linear-gradient(to bottom right,#e0e0e0);
      opacity: .6;
    }
  }
  &-info {
    background-color: $white-color;
    border: 1px solid $outline-color;
    padding: 16px 0 16px 16px;
    width: auto;
    border-bottom: none;
    border-radius: 0 0 2px 2px;
    &__container {
      display: flex;
    }
  }
  &__image {
    height: 100%;
    width: 100%;
    background-color: #DCDCDC;
    object-fit: cover;
  }
  &__flag {
    height: 36px;
    width: 36px;
    border-radius: 50%;
    background-position: center center;
    margin-right: 12px;
  }
  &__name {
    color: $text-highlight-color;
    text-transform: uppercase;
    font-size: 16px;
    font-family: $bold-font;
    line-height: 21px;
    margin-bottom: 5px;
  }
  &__location {
    color: $text-color;
    font-family: $normal-font;
    font-size: 16px;
  }
  &__description {
    padding: 0 16px 21px 16px;
    background-color: $white-color;
    border: 1px solid $outline-color;
    border-top: none;
    border-radius: 0 0 2px 2px;
  }
  &__facility {
    display: flex;
    justify-content: space-between;
    margin-top: 17px;
    &-number {
      font-family: $bold-font;
      color: $text-highlight-color;
      margin-right: 6px;
    }
  }
  &__icon {
    margin-right: 5px;
    &-container {
      display: flex;
      align-items: center;
      flex-wrap: wrap;
      color: $text-color;
      font-family: $normal-font;
      font-size: 16px;
      line-height: 21px;
    }
  }
  &__horizontal-line {
    width: inherit;
    box-sizing: border-box;
    height: 1px;
    background: $outline-color;
  }
}

.restore-acc-btn {
  padding: 5px 20px;
  cursor: pointer;
  outline: none;
  top: 50%;
  left: 50%;
  margin-left: -50px;
  margin-top: -30px;
  font-size: 1rem !important;

  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  color: #fff;
  display: inline-block;
  text-align: center;
  background-color: #3359DB;
  position: absolute;
}

.restore-acc-btn:hover {
  background-color: rgb(32, 64, 167);
}
