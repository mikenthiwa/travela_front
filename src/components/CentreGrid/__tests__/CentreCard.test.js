import React from 'react';
import { shallow } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import CentreCard from '../CentreCard';

const props = {
  cardImage: 'https:someimage.png',
  imageAlt: 'an image',
  countryFlagImage: 'https://countryimage.png',
  guestHouseName: 'Guest House',
  guestHouseLocation: 'Kigali, Rwanda',
  beds: 5,
  bathrooms: 4,
  handleOnRestore: jest.fn(),
  guesthouseId: 'guest-house-1',
  disabledGuestHouse: {
    id: 'guest-house-1',
  },
  guestHouse: {
    genderPolicy: 'male'
  }
};

describe('<CentreCard />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CentreCard {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('displays the correct data', () => {
    expect(wrapper.find('.centre__name').text()).toEqual('Guest House');
    expect(wrapper.find('.centre__location').text()).toEqual('Kigali, Rwanda');
    expect(wrapper.find('.centre__name').text()).toEqual('Guest House');
    expect(wrapper.find('.centre__icon-container').at(0).text()).toEqual('5Beds');
    expect(wrapper.find('.centre__icon-container').at(1).text()).toEqual('4Bathrooms');
  });

  it('renders the correct image', () => {
    expect(wrapper.find('.centre__image').prop('src')).toEqual('https:someimage.png');
    expect(wrapper.find('.centre__flag').prop('style')).toEqual({
      backgroundImage: 'url(https://countryimage.png)'
    });
    expect(wrapper.find('.centre__flag').prop('alt')).toEqual('Kigali, Rwanda flag');
  });

  it('should call preventDefault when the button is clicked', () => {
    const newProps = {
      ...props,
      disabledGuestHouse: {}
    };

    const mockedEvent = { preventDefault: jest.fn() };
    wrapper = shallow(<CentreCard {...newProps} />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'handleClick');

    const link = wrapper.find('#thisGuesthouseLink');
    link.simulate('click', mockedEvent);
    expect(instance.handleClick).toBeCalled;
  });

  it('should call handleOnRestore when the button is clicked', () => {
    wrapper = mount(
      <Router>
        <CentreCard {...props} />
      </Router>
    );
    const { handleOnRestore } = props;
    const button = wrapper.find('.restore-acc-btn');
    button.simulate('click');
    expect(handleOnRestore).toBeCalled;
  });
});
