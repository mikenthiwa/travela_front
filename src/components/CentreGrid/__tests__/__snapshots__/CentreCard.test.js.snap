// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CentreCard /> renders correctly 1`] = `
<div
  className="mdl-cell mdl-cell--4 mdl-card centre-card-disabled"
>
  <Link
    id="thisGuesthouseLink"
    onClick={[Function]}
    replace={false}
    to="/residence/manage/guest-houses/guest-house-1"
  >
    <div
      className="mdl-card__media centre-card__image"
    >
      <img
        alt="an image"
        className="centre__image"
        onError={[Function]}
        src="https:someimage.png"
      />
    </div>
    <Button
      altText=""
      badge={null}
      badgeClass=""
      buttonClass="restore-acc-btn"
      buttonId=""
      disabled={false}
      imageClass="mdl-Icon"
      imageSrc=""
      onClick={[Function]}
      responsiveText=""
      reverseText={false}
      showBadge={false}
      text="Restore"
      textClass=""
    />
    <div
      className="mdl-card__supporting-text centre-info"
    >
      <div
        className="centre-info__container"
      >
        <div
          alt="Kigali, Rwanda flag"
          className="centre__flag"
          style={
            Object {
              "backgroundImage": "url(https://countryimage.png)",
            }
          }
        />
        <div>
          <p
            className="centre__name"
          >
            Guest House
          </p>
          <span
            className="centre__location"
          >
            Kigali, Rwanda
          </span>
        </div>
      </div>
    </div>
    <div
      className="mdl-card__actions centre__description"
    >
      <div
        className="centre__horizontal-line"
      />
      <div
        className="centre__facility"
      >
        <div
          className="centre__icon-container"
        >
          <img
            alt="Bed number"
            className="centre__icon"
            src="accomodation-grey.svg"
          />
          <span
            className="centre__facility-number"
          >
            5
          </span>
          Beds
        </div>
        <img
          alt="male"
          src="undefined"
          title="male"
        />
        <div
          className="centre__icon-container"
        >
          <img
            alt="Bathroom number"
            className="centre__icon"
            src="hot_tub_24px.svg"
          />
          <span
            className="centre__facility-number"
          >
            4
          </span>
          Bathrooms
        </div>
      </div>
    </div>
  </Link>
</div>
`;
